src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -0,0 +1,280 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc.Razor.Extensions;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.NET.Sdk.Razor.SourceGenerators
+{
+    [Generator]
+    public partial class RazorSourceGenerator : ISourceGenerator
+    {
+        private static readonly ParallelOptions DefaultParallelOptions = new();
+
+        public void Initialize(GeneratorInitializationContext context)
+        {
+        }
+
+        public void Execute(GeneratorExecutionContext context)
+        {
+            var razorContext = RazorSourceGenerationContext.Create(context);
+            if (razorContext is null ||
+                (razorContext.RazorFiles.Count == 0 && razorContext.CshtmlFiles.Count == 0))
+            {
+                return;
+            }
+
+            HandleDebugSwitch(razorContext.WaitForDebugger);
+
+            var tagHelpers = ResolveTagHelperDescriptors(context, razorContext);
+
+            var projectEngine = RazorProjectEngine.Create(razorContext.Configuration, razorContext.FileSystem, b =>
+            {
+                b.Features.Add(new DefaultTypeNameFeature());
+                b.SetRootNamespace(razorContext.RootNamespace);
+
+                b.Features.Add(new StaticTagHelperFeature { TagHelpers = tagHelpers, });
+                b.Features.Add(new DefaultTagHelperDescriptorProvider());
+
+                CompilerFeatures.Register(b);
+                RazorExtensions.Register(b);
+
+                b.SetCSharpLanguageVersion(((CSharpParseOptions)context.ParseOptions).LanguageVersion);
+            });
+
+            CodeGenerateRazorComponents(context, razorContext, projectEngine);
+            GenerateViews(context, razorContext, projectEngine);
+        }
+
+        private void GenerateViews(GeneratorExecutionContext context, RazorSourceGenerationContext razorContext, RazorProjectEngine projectEngine)
+        {
+            var files = razorContext.CshtmlFiles;
+
+            Parallel.For(0, files.Count, GetParallelOptions(), i =>
```

Comment:
Please make sure to pass down `context.CancellationToken` for parallelization work like this.